source/4_tools/regina/src/regina.cpp

```diff
@@ -1,54 +1,865 @@
+        auto grid = mesh_factory::create_grid<u16>(20.0f, 30.0f, 60, 40);
+        auto sphere = mesh_factory::create_sphere<u16>(0.5f, 20, 20);
+        auto cylinder = mesh_factory::create_cylinder<u16>(0.5f, 0.3f, 3.0f, 20, 20);
+
+        //
+        // We are concatenating all the geometry into one big vertex/index buffer.  So
+        // define the regions in the buffer each submesh covers.
+        //
+
+        // Cache the vertex offsets to each object in the concatenated vertex buffer.
+        u32 box_vertex_offset = 0;
+        u32 grid_vertex_offset = box.vertices().size();
+        u32 sphere_verteX_offset = grid_vertex_offset + grid.vertices().size();
+        u32 cylinder_vertex_offset = sphere_verteX_offset + sphere.vertices().size();
+
+        // Cache the starting index for each object in the concatenated index buffer.
+        u32 box_index_offset = 0;
+        u32 grid_index_offset = box.indices().size();
+        u32 sphere_index_offset = grid_index_offset + grid.indices().size();
+        u32 cylinder_index_offset = sphere_index_offset + sphere.indices().size();
+
+        // Define the SubmeshGeometry that cover different 
+        // regions of the vertex/index buffers.
+
+        renderer::Submesh box_submesh;
+        box_submesh.index_count = box.indices().size();
+        box_submesh.start_index_location = box_index_offset;
+        box_submesh.base_vertex_location = box_vertex_offset;
+
+        renderer::Submesh grid_submesh;
+        grid_submesh.index_count = grid.indices().size();
+        grid_submesh.start_index_location = grid_index_offset;
+        grid_submesh.base_vertex_location = grid_vertex_offset;
+
+        renderer::Submesh sphere_submesh;
+        sphere_submesh.index_count = sphere.indices().size();
+        sphere_submesh.start_index_location = sphere_index_offset;
+        sphere_submesh.base_vertex_location = sphere_verteX_offset;
+
+        renderer::Submesh cylinder_submesh;
+        cylinder_submesh.index_count = cylinder.indices().size();
+        cylinder_submesh.start_index_location = cylinder_index_offset;
+        cylinder_submesh.base_vertex_location = cylinder_vertex_offset;
+
+        //
+        // Extract the vertex elements we are interested in and pack the
+        // vertices of all the meshes into one vertex buffer.
+        //
+
+        auto total_vertex_count = box.vertices().size() + grid.vertices().size() + sphere.vertices().size() + cylinder.vertices().size();
+        auto total_index_count = box.indices().size() + grid.indices().size() + sphere.indices().size() + cylinder.indices().size();
+
+        rsl::vector<renderer::VertexPosCol> vertices((rsl::Size)total_vertex_count);
+
+        u32 k = 0;
+        for (s32 i = 0; i < box.vertices().size(); ++i, ++k)
+        {
+            glm::vec3 position = box.vertices()[i].position;
+            glm::vec4 color = { 1.000000000f, 0.270588249f, 0.000000000f, 1.000000000f }; // Orange Red
```

Comment:
we have predefined colors, please use them

Reply:
Those are within a different format than required here, I will either need a conversion step or the colors need to be predefined differently.

Reply:
not necessarily, you just need a memcpy

Reply:
I think there's an even better way to writing this, I'll make an issue for it

Reply:
linked to https://github.com/New-Creta/Rex/issues/266